src/Nethermind/Nethermind.Serialization.Rlp/Rlp.cs

```diff
@@ -73,10 +75,32 @@ public Rlp(byte[] bytes)
 
         public int Length => Bytes.Length;
 
-        public static readonly Dictionary<Type, IRlpDecoder> Decoders = new();
+        public static FrozenDictionary<TypeAsKey, IRlpDecoder> Decoders = new Dictionary<TypeAsKey, IRlpDecoder>().ToFrozenDictionary();
+
+        public struct TypeAsKey(Type key) : IEquatable<TypeAsKey>
+        {
+            private readonly Type _key = key;
+            public Type Value => _key;
+
+            public static implicit operator Type(TypeAsKey key) => key._key;
+            public static implicit operator TypeAsKey(Type key) => new(key);
+
+            public bool Equals(TypeAsKey other) => _key.Equals(other._key);
+            public override int GetHashCode() => _key.GetHashCode();
+        }
+
+        public static void RegisterDecoder(Type type, IRlpDecoder decoder)
+        {
+            Dictionary<TypeAsKey, IRlpDecoder> dict = new(Decoders)
+            {
+                [type] = decoder
+            };
+            Decoders = dict.ToFrozenDictionary();
```

Comment:
should we eagerly create frozen dictionary or better create it lazy?

Reply:
Something like? https://github.com/NethermindEth/nethermind/pull/6957/commits/052d8b0982dceff711f8a2533e08447cd66ab9e4